From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Istvan Zoltan Ballok <istvan.zoltan.ballok@sap.com>
Date: Tue, 10 Dec 2024 13:45:58 +0100
Subject: [PATCH 05/29] NewCooperativeClient wraps controller-runtime/Client

NewCooperativeClient wraps a client of type
sigs.k8s.io/controller-runtime/pkg/client/Client.

The Cooperative Client can be used together with the Cooperative Executor to
detect possible race conditions in functions that interact with the Kubernetes
Client API.
---
 .../race/cooperative_client.go                | 191 ++++++++++++++
 .../race/cooperative_client_test.go           | 244 ++++++++++++++++++
 2 files changed, 435 insertions(+)
 create mode 100644 pkg/resourcemanager/controller/garbagecollector/race/cooperative_client.go
 create mode 100644 pkg/resourcemanager/controller/garbagecollector/race/cooperative_client_test.go

diff --git a/pkg/resourcemanager/controller/garbagecollector/race/cooperative_client.go b/pkg/resourcemanager/controller/garbagecollector/race/cooperative_client.go
new file mode 100644
index 0000000000..e95a7edaab
--- /dev/null
+++ b/pkg/resourcemanager/controller/garbagecollector/race/cooperative_client.go
@@ -0,0 +1,191 @@
+// SPDX-FileCopyrightText: 2025 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package race
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	resourcesv1alpha1 "github.com/gardener/gardener/pkg/apis/resources/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func NewCooperativeClient(clientName string, client client.Client, executor CooperativeExecutor) client.Client {
+	return &cooperativeClient{
+		clientName: clientName,
+		client:     client,
+		executor:   executor,
+	}
+}
+
+type cooperativeClient struct {
+	clientName string
+	client     client.Client
+	executor   CooperativeExecutor
+	mu         sync.Mutex
+	callID     int
+}
+
+func (cc *cooperativeClient) nextCallID() int {
+	cc.mu.Lock()
+	defer cc.mu.Unlock()
+	callID := cc.callID
+	cc.callID++
+	return callID
+}
+
+func (cc *cooperativeClient) Create(ctx context.Context, obj client.Object, opts ...client.CreateOption) error {
+	callID := cc.nextCallID()
+	cc.executor.LockAndYield(cc.clientName, callID, fmt.Sprintf("Create(%v, %v)", ObjectToString(obj), opts))
+	err := cc.client.Create(ctx, obj, opts...)
+	cc.executor.Release(cc.clientName, callID, fmt.Sprintf("object=%s, error=%v", ObjectToString(obj), err))
+	return err
+}
+
+func (cc *cooperativeClient) Get(ctx context.Context, key client.ObjectKey, obj client.Object, opts ...client.GetOption) error {
+	if _, ok := obj.(*corev1.Secret); ok {
+		callID := cc.nextCallID()
+		cc.executor.LockAndYield(cc.clientName, callID, fmt.Sprintf("Get(%v, %v)", key, opts))
+		err := cc.client.Get(ctx, key, obj, opts...)
+		cc.executor.Release(cc.clientName, callID, fmt.Sprintf("object=%s, error=%v", ObjectToString(obj), err))
+		return err
+	}
+	return cc.client.Get(ctx, key, obj, opts...)
+}
+
+func (cc *cooperativeClient) List(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
+	if list.GetObjectKind().GroupVersionKind().Kind == "SecretList" ||
+		list.GetObjectKind().GroupVersionKind().Kind == "ManagedResourceList" {
+		callID := cc.nextCallID()
+		cc.executor.LockAndYield(cc.clientName, callID, fmt.Sprintf("List(%s, %v)", list.GetObjectKind().GroupVersionKind().Kind, opts))
+		err := cc.client.List(ctx, list, opts...)
+		cc.executor.Release(cc.clientName, callID, fmt.Sprintf("list=%v, error=%v", ListToString(list), err))
+		return err
+	}
+	return cc.client.List(ctx, list, opts...)
+}
+
+func (cc *cooperativeClient) Patch(ctx context.Context, obj client.Object, patch client.Patch, opts ...client.PatchOption) error {
+	callID := cc.nextCallID()
+	cc.executor.LockAndYield(cc.clientName, callID, fmt.Sprintf("Patch(%v, %v, %v)", ObjectToString(obj), PatchToString(obj, patch), opts))
+	err := cc.client.Patch(ctx, obj, patch, opts...)
+	cc.executor.Release(cc.clientName, callID, fmt.Sprintf("object=%v, error=%v", ObjectToString(obj), err))
+	return err
+}
+
+func (cc *cooperativeClient) Update(ctx context.Context, obj client.Object, opts ...client.UpdateOption) error {
+	callID := cc.nextCallID()
+	cc.executor.LockAndYield(cc.clientName, callID, fmt.Sprintf("Update(%v, %v)", ObjectToString(obj), opts))
+	err := cc.client.Update(ctx, obj, opts...)
+	cc.executor.Release(cc.clientName, callID, fmt.Sprintf("object=%s, error=%v", ObjectToString(obj), err))
+	return err
+}
+
+func (cc *cooperativeClient) Delete(ctx context.Context, obj client.Object, opts ...client.DeleteOption) error {
+	callID := cc.nextCallID()
+	cc.executor.LockAndYield(cc.clientName, callID, fmt.Sprintf("Delete(%v, %s)", ObjectToString(obj), DeleteOptionToString(opts)))
+	err := cc.client.Delete(ctx, obj, opts...)
+	cc.executor.Release(cc.clientName, callID, fmt.Sprintf("error=%v", err))
+	return err
+}
+
+func (cc *cooperativeClient) Scheme() *runtime.Scheme     { return cc.client.Scheme() }
+func (cc *cooperativeClient) RESTMapper() meta.RESTMapper { return cc.client.RESTMapper() }
+func (cc *cooperativeClient) GroupVersionKindFor(obj runtime.Object) (schema.GroupVersionKind, error) {
+	return cc.client.GroupVersionKindFor(obj)
+}
+func (cc *cooperativeClient) IsObjectNamespaced(obj runtime.Object) (bool, error) {
+	return cc.client.IsObjectNamespaced(obj)
+}
+func (cc *cooperativeClient) DeleteAllOf(ctx context.Context, obj client.Object, opts ...client.DeleteAllOfOption) error {
+	return cc.client.DeleteAllOf(ctx, obj, opts...)
+}
+func (cc *cooperativeClient) Status() client.SubResourceWriter { return cc.client.Status() }
+func (cc *cooperativeClient) SubResource(subResource string) client.SubResourceClient {
+	return cc.client.SubResource(subResource)
+}
+
+func ObjectToString(obj client.Object) string {
+	fromString := func(name, value string) string {
+		if value == "" {
+			return ""
+		}
+		return fmt.Sprintf("%s: %s, ", name, value)
+	}
+	fromTime := func(name string, value metav1.Time) string {
+		if value.IsZero() {
+			return ""
+		}
+		return fmt.Sprintf("%s: %s, ", name, value)
+	}
+	fromMap := func(name string, value map[string]string) string {
+		if len(value) == 0 {
+			return ""
+		}
+		return fmt.Sprintf("%s: %v, ", name, value)
+	}
+	fromBool := func(name string, value *bool) string {
+		if value == nil {
+			return ""
+		}
+		return fmt.Sprintf("%s: %t, ", name, *value)
+	}
+	objectMetaToString := func(obj metav1.Object) string {
+		result := fromString("Namespace", obj.GetNamespace())
+		result += fromString("Name", obj.GetName())
+		result += fromString("ResourceVersion", obj.GetResourceVersion())
+		result += fromTime("CreationTimestamp", obj.GetCreationTimestamp())
+		result += fromMap("Labels", obj.GetLabels())
+		result += fromMap("Annotations", obj.GetAnnotations())
+		return result
+	}
+	switch obj := obj.(type) {
+	case *corev1.Secret:
+		result := objectMetaToString(obj)
+		result += fromBool("Immutable", obj.Immutable)
+		return fmt.Sprintf("&Secret{%s}", result)
+	case *resourcesv1alpha1.ManagedResource:
+		result := objectMetaToString(obj)
+		return fmt.Sprintf("&ManagedResource{%s}", result)
+	case *metav1.PartialObjectMetadata:
+		result := objectMetaToString(obj)
+		return fmt.Sprintf("&PartialObjectMetadata{%s}", result)
+	}
+	return fmt.Sprintf("%v", obj)
+}
+
+func ListToString(list client.ObjectList) string {
+	if list, ok := list.(*metav1.PartialObjectMetadataList); ok {
+		result := ""
+		for _, item := range list.Items {
+			result += fmt.Sprintf("%s, ", ObjectToString(&item))
+		}
+		return fmt.Sprintf("&PartialObjectMetadataList{Items: [%s]}", result)
+	}
+	return fmt.Sprintf("%v", list)
+}
+
+func DeleteOptionToString(opts []client.DeleteOption) string {
+	if len(opts) != 1 {
+		return fmt.Sprintf("%v", opts)
+	}
+	if p, ok := opts[0].(client.Preconditions); ok {
+		if p.UID != nil && p.ResourceVersion != nil {
+			return fmt.Sprintf("Preconditions{UID: %v, ResourceVersion: %v}", *p.UID, *p.ResourceVersion)
+		}
+	}
+	return ""
+}
+
+func PatchToString(obj client.Object, patch client.Patch) string {
+	data, _ := patch.Data(obj)
+	return fmt.Sprintf("Patch{%v, %s}", patch.Type(), data)
+}
diff --git a/pkg/resourcemanager/controller/garbagecollector/race/cooperative_client_test.go b/pkg/resourcemanager/controller/garbagecollector/race/cooperative_client_test.go
new file mode 100644
index 0000000000..eb9a9669b9
--- /dev/null
+++ b/pkg/resourcemanager/controller/garbagecollector/race/cooperative_client_test.go
@@ -0,0 +1,244 @@
+// SPDX-FileCopyrightText: 2025 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package race_test
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+
+	resourcesv1alpha1 "github.com/gardener/gardener/pkg/apis/resources/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/resourcemanager/controller/garbagecollector/race"
+	kubernetesutils "github.com/gardener/gardener/pkg/utils/kubernetes"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	fakeclient "sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+var _ = Describe("Cooperative Client", func() {
+	const (
+		LABEL_KEY = "resources.gardener.cloud/foo"
+	)
+	var (
+		ctx        = context.TODO()
+		executor   race.CooperativeExecutor
+		c          client.Client
+		cA, cB     client.Client
+		secretName string
+		setup      func()
+		assert     func() error
+
+		getSecret = func(secretName string) *corev1.Secret {
+			return &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      secretName,
+					Namespace: metav1.NamespaceDefault,
+				},
+			}
+		}
+		getValue = func(secret *corev1.Secret) int {
+			value, _ := strconv.Atoi(secret.Labels[LABEL_KEY])
+			return value
+		}
+		setValue = func(secret *corev1.Secret, value int) {
+			secret.Labels[LABEL_KEY] = strconv.Itoa(value)
+		}
+	)
+
+	BeforeEach(func() {
+		executor = race.NewCooperativeExecutor(race.DEBUG)
+		setup = func() {
+			c = fakeclient.NewClientBuilder().WithScheme(kubernetes.SeedScheme).Build()
+			cA = race.NewCooperativeClient("A", c, executor)
+			cB = race.NewCooperativeClient("B", c, executor)
+
+			secret := getSecret("secret")
+			kubernetesutils.MakeUnique(secret)
+			secretName = secret.Name
+			c.Create(ctx, secret)
+		}
+		assert = func() error {
+			secret := getSecret(secretName)
+			c.Get(ctx, client.ObjectKeyFromObject(secret), secret)
+			if getValue(secret) != 2 {
+				return fmt.Errorf("Assertion failed: expected 2, got %d", getValue(secret))
+			}
+			return nil
+		}
+	})
+
+	It("should detect possible race conditions when actors do not properly use the optimistic locking of the Kubernetes client", func() {
+		result := executor.RunAllCombinations(
+			setup,
+			assert,
+			func() {
+				secret := getSecret(secretName)
+				mutate := func() error {
+					setValue(secret, getValue(secret)+1)
+					return nil
+				}
+				controllerutil.CreateOrUpdate(ctx, cA, secret, mutate)
+			},
+			func() {
+				secret := getSecret(secretName)
+				mutate := func() error {
+					setValue(secret, getValue(secret)+1)
+					return nil
+				}
+				controllerutil.CreateOrUpdate(ctx, cB, secret, mutate)
+			},
+		)
+		Expect(result.NumberOfPathsWithFailedAssertion).To(Equal(4))
+		Expect(result.NumberOfPathsChecked).To(Equal(6))
+		Expect(strings.TrimSpace(result.PathsChecked)).To(Equal(strings.TrimSpace(`
+
+Path 1:
+- A/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- A/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- B/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- B/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:2 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 3, Labels: map[resources.gardener.cloud/foo:2 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+  assertion error: <nil>
+Path 2:
+- B/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- A/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- A/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- B/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=Operation cannot be fulfilled on secrets "secret-e3b0c442": object was modified
+  assertion error: Assertion failed: expected 2, got 1
+Path 3:
+- A/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- B/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- A/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- B/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=Operation cannot be fulfilled on secrets "secret-e3b0c442": object was modified
+  assertion error: Assertion failed: expected 2, got 1
+Path 4:
+- B/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- B/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- A/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- A/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:2 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 3, Labels: map[resources.gardener.cloud/foo:2 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+  assertion error: <nil>
+Path 5:
+- B/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- A/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- B/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- A/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=Operation cannot be fulfilled on secrets "secret-e3b0c442": object was modified
+  assertion error: Assertion failed: expected 2, got 1
+Path 6:
+- A/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- B/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- B/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- A/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=Operation cannot be fulfilled on secrets "secret-e3b0c442": object was modified
+  assertion error: Assertion failed: expected 2, got 1
+
+		`)), "Here is the actual value of PathsChecked in full length. Copy this and update the string literal in the unit test's source code: \n\n%s\n",
+			result.PathsChecked)
+	})
+
+	It("should detect no race conditions when actors properly use the optimistic locking of the Kubernetes client", func() {
+		result := executor.RunAllCombinations(
+			setup,
+			assert,
+			func() {
+				secret := getSecret(secretName)
+				mutate := func() error {
+					setValue(secret, getValue(secret)+1)
+					return nil
+				}
+				runs := 0
+				for {
+					if _, err := controllerutil.CreateOrUpdate(ctx, cA, secret, mutate); err == nil {
+						runs++
+						break
+					}
+				}
+				Expect(runs).To(BeNumerically("<=", 2))
+			},
+			func() {
+				secret := getSecret(secretName)
+				mutate := func() error {
+					setValue(secret, getValue(secret)+1)
+					return nil
+				}
+				runs := 0
+				for {
+					if _, err := controllerutil.CreateOrUpdate(ctx, cB, secret, mutate); err == nil {
+						runs++
+						break
+					}
+				}
+				Expect(runs).To(BeNumerically("<=", 2))
+			},
+		)
+		Expect(result.NumberOfPathsWithFailedAssertion).To(Equal(0))
+		Expect(result.NumberOfPathsChecked).To(Equal(6))
+		Expect(strings.TrimSpace(result.PathsChecked)).To(Equal(strings.TrimSpace(`
+
+Path 1:
+- A/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- A/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- B/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- B/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:2 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 3, Labels: map[resources.gardener.cloud/foo:2 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+  assertion error: <nil>
+Path 2:
+- B/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- A/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- A/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- B/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=Operation cannot be fulfilled on secrets "secret-e3b0c442": object was modified
+- B/2 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- B/3 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:2 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 3, Labels: map[resources.gardener.cloud/foo:2 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+  assertion error: <nil>
+Path 3:
+- A/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- B/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- A/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- B/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=Operation cannot be fulfilled on secrets "secret-e3b0c442": object was modified
+- B/2 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- B/3 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:2 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 3, Labels: map[resources.gardener.cloud/foo:2 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+  assertion error: <nil>
+Path 4:
+- B/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- B/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- A/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- A/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:2 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 3, Labels: map[resources.gardener.cloud/foo:2 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+  assertion error: <nil>
+Path 5:
+- B/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- A/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- B/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- A/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=Operation cannot be fulfilled on secrets "secret-e3b0c442": object was modified
+- A/2 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- A/3 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:2 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 3, Labels: map[resources.gardener.cloud/foo:2 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+  assertion error: <nil>
+Path 6:
+- A/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- B/0 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- B/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- A/1 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 1, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=Operation cannot be fulfilled on secrets "secret-e3b0c442": object was modified
+- A/2 Get(default/secret-e3b0c442, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:1 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+- A/3 Update(&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 2, Labels: map[resources.gardener.cloud/foo:2 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, []) -> object=&Secret{Namespace: default, Name: secret-e3b0c442, ResourceVersion: 3, Labels: map[resources.gardener.cloud/foo:2 resources.gardener.cloud/garbage-collectable-reference:true resources.gardener.cloud/used:true], Immutable: true, }, error=<nil>
+  assertion error: <nil>
+
+		`)), "Here is the actual value of PathsChecked in full length. Copy this and update the string literal in the unit test's source code: \n\n%s\n",
+			result.PathsChecked)
+	})
+
+	It("should concisely print objects", func() {
+		Expect(race.ObjectToString(&corev1.Secret{ObjectMeta: metav1.ObjectMeta{
+			Name:      "secret",
+			Namespace: metav1.NamespaceDefault,
+		}})).To(Equal(
+			"&Secret{Namespace: default, Name: secret, }"))
+		Expect(race.ObjectToString(&resourcesv1alpha1.ManagedResource{ObjectMeta: metav1.ObjectMeta{
+			Name:      "mr",
+			Namespace: metav1.NamespaceDefault,
+		}})).To(Equal(
+			"&ManagedResource{Namespace: default, Name: mr, }"))
+	})
+})
